File: fix_migration_ids.py
import re

# Path to the file
file_path = r"c:\Users\zakio\Documents\Project\PB-Next\pb_migrations\1766600000_batch3_kits_seed.js.bak"

with open(file_path, 'r', encoding='utf-8') as f:
    content = f.read()

# Find all "id": "value" occurrences to identify short IDs
# We look for IDs that are between 1 and 14 chars.
# We mostly care about the ones in the definitions (BATCH3_ITEMS, KIT_TEMPLATES, KIT_ITEMS)
# But strictly speaking, any ID assignment that is short is invalid for creation.

# Regex to find definition IDs. 
# Matches: "id": "short_id"
id_pattern = re.compile(r'"id":\s*"([a-zA-Z0-9_]{1,14})"')

matches = set(id_pattern.findall(content))

replacements = {}

print("Found short IDs:")
for short_id in matches:
    # Generate replacement: pad with 'x' to 15 chars
    # We use 'x' suffix to keep it readable
    needed = 15 - len(short_id)
    new_id = short_id + ('x' * needed)
    replacements[short_id] = new_id
    print(f"  {short_id} -> {new_id}")

# Replace in content
# We must be careful not to replace partial matches if an ID is a substring of another.
# But IDs here are quoted strings. So replacing "old_id" with "new_id" is safe.
# We also need to replace references like "template_id": "short_id" or "main_product_id": "short_id"

new_content = content

# Sort replacements by length descending to avoid substring collision issues (e.g. replacing 'id1' inside 'id10')
sorted_keys = sorted(replacements.keys(), key=len, reverse=True)

for old_id in sorted_keys:
    new_id = replacements[old_id]
    # Replace quoted occurrences
    # This covers "id": "old", "template_id": "old", "value": "old" etc.
    new_content = new_content.replace(f'"{old_id}"', f'"{new_id}"')

# Write back to original filename (remove .bak)
target_path = file_path.replace('.bak', '')
with open(target_path, 'w', encoding='utf-8') as f:
    f.write(new_content)

print(f"Fixed file written to {target_path}")

File: fix_migration_ids.js
const fs = require('fs');
const path = require('path');

// Target the .js file now since we overwrote it previously, or fallback to .bak if needed.
// User output shows: migration 1766600000_batch3_kits_seed.js
const filePath = String.raw`c:\Users\zakio\Documents\Project\PB-Next\pb_migrations\1766600000_batch3_kits_seed.js`;

try {
    let content = fs.readFileSync(filePath, 'utf8');

    // We need to find ALL manually assigned IDs that are being used.
    // The previous regex caught them. We can reuse the logic but improve the "newId" generation.

    // Regex to find "id": "value"
    // We capture the ID value.
    // We relax the regex to catch underscores and potentially longer things we messed up.
    // But primarily we want to catch the ones we just padded or the original ones if we are running on clean state.
    // Since we ran the script once, we might have IDs like "kit_venice_pkgx". 
    // We want to turn "kit_venice_pkgx" -> "kitvenicepkgxxx" (15 chars, no underscore)

    // Let's sweep for any ID-like string that is in our dataset.
    // The dataset keys are: id, main_product_id, template_id, product_id
    // But simpler approach: Find anything looking like a manual ID string.
    // Our manual IDs usually start with: lens_, kit_, cam_, mb_, supp_, wl_, stab_, item_, mon_
    // or the 'x' padded versions from previous run.

    const idRegex = /"(id|main_product_id|template_id|product_id)":\s*"([^"]+)"/g;

    const candidates = new Set();
    let match;
    while ((match = idRegex.exec(content)) !== null) {
        // match[2] is the value
        const val = match[2];
        // Filter out system IDs (usually 15 mixed chars, hard to distinguish from our bad ones if we padded them)
        // But our manual ones have meaningful prefixes usually or underscores.
        if (val.includes('_') || val.length !== 15 || /[^a-z0-9]/.test(val)) {
            candidates.add(val);
        }
    }

    console.log(`Found ${candidates.size} candidate invalid ID strings.`);

    const replacements = {};

    candidates.forEach(oldId => {
        // Sanitize: remove non-alphanumeric, toLowerCase
        let stripped = oldId.replace(/[^a-z0-9]/gi, '').toLowerCase();

        let newId = stripped;
        if (newId.length < 15) {
            newId = newId + 'x'.repeat(15 - newId.length);
        } else if (newId.length > 15) {
            newId = newId.substring(0, 15);
        }

        if (oldId !== newId) {
            replacements[oldId] = newId;
            console.log(`  "${oldId}" -> "${newId}"`);
        }
    });

    // Replace in content
    // Sort by length desc to avoid substring issues
    const sortedKeys = Object.keys(replacements).sort((a, b) => b.length - a.length);

    for (const oldId of sortedKeys) {
        const newId = replacements[oldId];
        // Replace all quoted occurrences globally
        // This is safe because these specific semantic strings are unique enough in this context
        // and we are matching exact quoted strings mostly by the way we found them (though we do global replace)
        const globalRegex = new RegExp(`"${oldId}"`, 'g');
        content = content.replace(globalRegex, `"${newId}"`);
    }

    fs.writeFileSync(filePath, content, 'utf8');
    console.log(`Fixed file written to ${filePath}`);

} catch (err) {
    console.error("Error:", err);
    process.exit(1);
}
